### ISSUE TITLE
[REVIEW]: OpenQuestion: A survey building and reporting platform written in Python

### ISSUE THREAD

**Issue by whedon** [openjournals/jose-reviews#123]
**Submitting author:** @Alcampopiano (<a href="http://orcid.org/0000-0002-3280-4447">Allan Campopiano</a>)
**Repository:** <a href="https://github.com/Alcampopiano/OpenQuestion" target ="_blank">https://github.com/Alcampopiano/OpenQuestion</a>
**Version:** v0.1.0
**Editor:** @labarba
**Reviewer:** @stefania11, @csev
**Archive:** Pending

**:warning: JOSE reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOES is currently operating in a "reduced service mode".

## Status

[![status](https://jose.theoj.org/papers/e0c00522459134c83676de25e2ccd253/status.svg)](https://jose.theoj.org/papers/e0c00522459134c83676de25e2ccd253)

Status badge code:

```
HTML: <a href="https://jose.theoj.org/papers/e0c00522459134c83676de25e2ccd253"><img src="https://jose.theoj.org/papers/e0c00522459134c83676de25e2ccd253/status.svg"></a>
Markdown: [![status](https://jose.theoj.org/papers/e0c00522459134c83676de25e2ccd253/status.svg)](https://jose.theoj.org/papers/e0c00522459134c83676de25e2ccd253)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/Alcampopiano/OpenQuestion" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@stefania11 & @csev, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/jose-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @labarba know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/jose/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/Alcampopiano/OpenQuestion">repository url</a>?
- [x ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Version:** Does the release version given match the GitHub release (v0.1.0)? - "No releases published"
- [x] **Authorship:** Has the submitting author (@Alcampopiano) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
-  The installation instructions are minimal, insufficient, and hard to find ( it seems that the project requires the anvil-app-server package to test installation. Authors link some external documentation for how to install this package but the links are confusing and one cannot test the functionality of this repo without spending extensive time installing the anvil server infrastructure
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] no due to lack of clear installation instructions 
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally, these should be handled with an automated package management solution.)
- [x] **Example usage:** Do the authors include examples of how to use the software?
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSE conflict of interest policy](https://github.com/openjournals/jose/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSE code of conduct](https://jose.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/Alcampopiano/OpenQuestion">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v0.1.0)?
- [x] **Authorship:** Has the submitting author (@Alcampopiano) made substantial contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation? (and documentation is sufficient?)
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly stated list of dependencies? (Ideally these should be handled with an automated package management solution.)
- [ ] **Example usage:** Do the authors include examples of how to use the software?
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [ ] **A statement of need:** Do the authors clearly state the need for this software and who the target audience is?
- [ ] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @stefania11, @csev it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/jose-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/jose-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 1001

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.14 s (1027.9 files/s, 69566.1 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
YAML                            57              9              2           3009
Python                          69           1036            737           2476
Markdown                        11            242              0           1000
CSS                              2            161            108            622
HTML                             1              7              0             78
TeX                              1              5              0             51
-------------------------------------------------------------------------------
SUM:                           141           1460            847           7236
-------------------------------------------------------------------------------


Statistical information for the repository 'f95df8a9209926c6adea706e' was
gathered on 2021/07/28.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
alcampopiano                    25           321             95            4.56
campopianoa@hcdsb.or           169          6402           2304           95.44

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
alcampopiano                224           69.8          2.3               11.61
campopianoa@hcdsb.or       4100           64.0          6.0                6.44
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/TVCG.2016.2599030 is OK

MISSING DOIs

- 10.1016/j.jbi.2008.08.010 may be a valid DOI for title: Research electronic data capture (REDCap)—a metadata-driven methodology and workflow process for providing translational research informatics support

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/jose-papers/jose.00123/jose.00123/10.21105.jose.00123.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/jose-papers/blob/jose.00123/jose.00123/10.21105.jose.00123.pdf) :page_facing_up: :point_left:

**Comment by labarba** [comment]
👋  @stefania11, @csev —  Here is the Review issue for "OpenQuestion: A survey building and reporting platform written in Python." The article proof is linked on the comment just above, and the submission repository is linked at the very top of this thread. You both have a reviewer checklist to guide you, review criteria are listed in our [docs](https://openjournals.readthedocs.io/en/jose/review_criteria.html) and you an ask me anything here!

You should feel free to open an issue in the target repository if needed, adding a link to this review issue to get the cross-link show up here.

**Comment by Alcampopiano** [comment]
Hello @stefania11 and @csev,

Thank you for reviewing this submission. I have sent a hosted version of the program (with the required login information) to your emails. This way you can hop in and try things out straight away if interested.

Take care,
Al

**Comment by whedon** [comment]
:wave: @csev, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @stefania11, please update us on how your review is going (this is an automated reminder).

**Comment by stefania11** [comment]
I cannot edit the above issue. I tried accepting the linked invitation but I think link is expired. @labarba would you mind re-sending invite?

**Comment by labarba** [comment]
@whedon re-invite @stefania11 as reviewer

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@stefania11 please accept the invite by clicking this link: https://github.com/openjournals/jose-reviews/invitations

**Comment by csev** [comment]
@Alcampopiano - I am catching up with this - I tried to install this with pip on a Mac and got "ERROR: Failed building wheel for setproctitle" - if there is a workaround that you know of, let me know.  I can spin up a docker and try again in a fresh linux but that will take some time.  I just want to run through the install process and get it to come up.

**Comment by labarba** [comment]
@stefania11 – Thanks for getting started with your review! Feel free to open issues in the target repository with any questions to the author (linking this review thread), or simply add your comments here.

**Comment by labarba** [comment]
@whedon remind @stefania11 in 40 days

**Comment by whedon** [comment]
Reminder set for @stefania11 in 40 days

**Comment by whedon** [comment]
:wave: @stefania11, please update us on how your review is going (this is an automated reminder).

**Comment by labarba** [comment]
Hi all, 
I have communicated with `@stefania11` via email regarding this review, and she has stated that, after her first review of this work some months ago, she has no more bandwidth for it now. I also note that she entered review comments by editing the checklist post at the top of this issue thread, noting the following under **Installation**:

> The installation instructions are minimal, insufficient, and hard to find ( it seems that the project requires the anvil-app-server package to test installation. Authors link some external documentation for how to install this package but the links are confusing and one cannot test the functionality of this repo without spending extensive time installing the anvil server infrastructure.

I also notice that @csev entered the following comment above:

> I tried to install this with pip on a Mac and got "ERROR: Failed building wheel for setproctitle" - if there is a workaround that you know of, let me know. I can spin up a docker and try again in a fresh linux but that will take some time. I just want to run through the install process and get it to come up.

Both reviewers therefor point to serious barriers for reuse of this software, while we did not have much engagement from the author. Therefore, I am making an editorial decision to reject this submission for publication in JOSE. I appreciate the author's effort, intention, and work, but we look to publish on software or lessons that are ready for reuse by others.

**Comment by labarba** [comment]
@whedon reject

**Comment by whedon** [comment]
Paper rejected.

**Comment by Alcampopiano** [comment]
Hello @labarba 👋,

**Comment by Alcampopiano** [comment]
Hi @labarba 👋,

Thank you for jumping in here. I would please like to request a bit of grace.

I have been following this submission very closely. Over covid things really slowed down on JOSE (and everywhere) and I was patiently waiting for the reviewers to complete their initial review. They were pinged multiple times without response (totally understandable) and so I was happy to wait for the initial review to be completed.

I honestly did not realize there was action required of me as the reviewers were asked to come back to the thread to finish up (but I should have been explicitly present and for that I apologize).

> hiya, @stefania11, @csev – could you pop in this review thread again? We’re hoping this can move now, as it’s been quite some time. Thank you! :pray:

Regarding the installation issues. I believe these were related to the M1 Mac and have now been resolved. To demonstrate this, I've recorded a Loom video below of me installing the software by following the installation instructions in the documentation.

https://www.loom.com/share/ccea5aff0c97413a9beedd58ed7424a8

Since I can demonstrate that the installation does work as indicated in the documentation, may I have a bit of grace on the communication issues? As I mentioned, I misunderstood that things were waiting exclusively on me. 

This project and review has my full attention but I misunderstood that a reject/accept was hanging in the balance, waiting on me. I think that is a reasonable assumption on my part but I should have been more explicit with you all.

Please let me know what you think.

**Comment by Alcampopiano** [comment]
@labarba Friendly ping about the message above.

**Comment by csev** [comment]
@Alcampopiano - I am happy to pick this back up - especially if the install process works.   I really don't like to say "yes" to software submissions that breaks immediately :) But if we can work through this and I can verify it installs from the repo and readme - I can be thumbs up.  Having a server to test is nice but not sufficient for me if the install from documentation / README seems to be broken.

**Comment by Alcampopiano** [comment]
@csev I really appreciate your message and agree completely. I'm confident that installation can be a smooth process.

In part, this is because OpenQuestion is built on top of [Anvil's open-source runtime engine](https://anvil.works/open-source) which powers thousands of real-world applications (open source and commercial). If we experience any issues, I'm certain that we can resolve them since we would not be the only ones affected.

Having said that, I must do my part to explain the instructions **very** clearly. I've recorded [a short video](https://www.loom.com/share/ccea5aff0c97413a9beedd58ed7424a8) above where I go through the installation [from the docs](https://alcampopiano.github.io/OpenQuestion/installation/) line by line. 

@csev If you are willing, could you try the following instructions again (with a clean environment)? Please let me know if I can help in any way. I will be very present.

For convenience, here are the steps from the docs:

```
# install from pip
pip install OpenQuestion

# clone the OpenQuestion project into your current working directory
git clone https://github.com/Alcampopiano/OpenQuestion.git

# open up a python shell to interact with app's backend
anvil-app-server --app OpenQuestion --shell

# add a new row to the Users table
new_user=app_tables.users.add_row(email='your_username@example.com', enabled=True)

# add an encrypted password
# these steps are not needed if using Google or Microsoft Authentication
import bcrypt
password_hash = bcrypt.hashpw(b'new password', bcrypt.gensalt(16))
new_user['password_hash'] = password_hash.decode()

# leave the Python shell (this is not in the docs but I will add it)
quit()

# launch the app itself!
anvil-app-server --app OpenQuestion
```

Sometimes there are eccentricities with local operating systems but on my mac M1 and Linux machine, these steps seem to work as expected. I will work to resolve any issues that come up.

Thank you again for your response Charles. I really appreciate it.

**Comment by labarba** [comment]
@Alcampopiano – Since I rejected this submission on the basis of the broken installation, you would have to make a new submission once you iron things out. As I noted [above](https://github.com/openjournals/jose-reviews/issues/123#issuecomment-12130194410, I communicated with reviewer `@stefania11` via email, and after the unhappy installation attempt, this reviewer is no longer willing / available to review. And that is what happens too with potential users if they run into hard installation troubles! The goal of a JOSE submission of for the software or teaching module to be reusable by others.

**Comment by Alcampopiano** [comment]
@labarba Thanks for responding. I strongly believe that the review that prompted the rejection was rather **unfair (and unclear)**.  I want to kindly bring this to your attention as I think it deserves addressed.

- In your opinion, do you think installation is too difficult (given [this short video](https://www.loom.com/share/ccea5aff0c97413a9beedd58ed7424a8) or [the steps listed below](https://github.com/openjournals/jose-reviews/issues/123#issuecomment-1224250795) from the docs)?

- May I please get your thoughts on what I believe was a rather unfair and unclear review? For example:

>  _installation is hard to find_

However, it's listed under "installation" in the docs ToC. This is not hard to find in my opinion. Where else should it be listed if not there?

<img width="1094" alt="Screen Shot 2022-08-23 at 12 48 19 PM" src="https://user-images.githubusercontent.com/4624246/186216832-3f2859d6-d177-4b52-bc28-005eb544b4b7.png">

> _Authors link some external documentation for how to install this package but the links are confusing_

This is not true or clearly stated. The installation instructions are completely self contained. A `pip` and a `git clone` are sufficient for installing everything needed and this is clearly stated in the docs. 

> _...without spending extensive time installing the anvil server infrastructure_

This dependency is installed automatically with `pip` in the first line of the installation instructions. I do not understand what the reviewer means by "extensive time installing". There's no rule against having a dependency of course 🙂.

In addition: (1) reviewer 1's comments were rather hidden above (not posted as an issue and I wasn't tagged). However, I take some responsibility for not noticing this. (2) The second reviewer is happy to review and this counts for something I think 🙂.

@labarba I really appreciate all the work that goes into this reviewing and I apologize for the misunderstandings on my part during the process. 

I kindly ask for your opinion on the review I received, especially given my response and demonstrations.
